src/entities/priceImpact/index.ts

```diff
@@ -237,6 +391,59 @@ export class PriceImpact {
         return PriceImpactAmount.fromDecimal(`${priceImpact}`);
     };
 
+    /**
+     * Calculate price impact on removing liquidity for nested pools.
+     *
+     * Note: is based on the premise that the price impact on removing liquidity
+     * for nested pools is the sum of the price impacts of each remove liquidity
+     * operation in the nested pools.
+     * @param input same input used in the corresponding remove liquidity nested operation
+     * @param nestedPoolState same nested pool state used in the corresponding remove liquidity nested operation
+     * @returns price impact amount
+     */
+    static removeLiquidityNested = async (
+        input: RemoveLiquidityNestedSingleTokenInput,
+        nestedPoolState: NestedPoolState,
+    ): Promise<PriceImpactAmount> => {
+        // inputs are being validated within RemoveLiquidity
+
+        // simulate removing liquidity to get amounts out
+        const removeLiquidityNested = new RemoveLiquidityNested();
+        const { bptAmountIn, amountsOut } = await removeLiquidityNested.query(
+            input,
+            nestedPoolState,
+        );
+
+        // simulate adding liquidity to get amounts in
+        const addLiquidityNested = new AddLiquidityNested();
+        const addLiquidityNestedInput: AddLiquidityNestedInput = {
+            accountAddress: input.accountAddress,
+            chainId: input.chainId,
+            rpcUrl: input.rpcUrl,
+            useNativeAssetAsWrappedAmountIn:
+                input.useNativeAssetAsWrappedAmountOut,
+            fromInternalBalance: input.toInternalBalance,
+            amountsIn: amountsOut.map((a) => a.toInputAmount()),
+        };
+        const { bptOut } = await addLiquidityNested.query(
+            addLiquidityNestedInput,
+            nestedPoolState,
+        );
+
+        // get relevant amounts for price impact calculation
+        const amountInitial = parseFloat(bptAmountIn.toSignificant());
+        const amountFinal = parseFloat(bptOut.toSignificant());
+
+        // calculate price impact using ABA method
+        const priceImpact = (amountInitial - amountFinal) / amountInitial / 2;
+        return PriceImpactAmount.fromDecimal(`${priceImpact}`);
+    };
+
+    /**
+     * Calculate price impact on single swap operations
+     * @param input same input used in the corresponding single swap operation
+     * @returns price impact amount
+     */
```

Comment:
`singleSwap` won't work with V3. Could construct Path and use `Swap` class which should expose the query for both V2 and V3.

Reply:
I'm wondering if I should remove method as a whole.
The initial goal was to have the swap class using this helper internally, but that doesn't really make much sense.
If we already have the swap class providing price impact for a swap, this feels a bit redundant.
What do you think?

Reply:
The Swap class no longer has a PI implementation because we removed the maths so this was quite nice. I did a follow up PR [here](https://github.com/balancer/b-sdk/pull/252) to show how it could be used.

Reply:
I'm actually thinking of a way to have Price Impact for swaps being handled externally (the same way as add/remove liquidity) so we don't need to worry about v2/v3 or single/batch swaps.
Ideally we'd provide it with a `SwapInput` and it would be able to build both path and reverse path, query amounts and return price impact.
I'll create a separate tasks so we don't forget to take a look at this in the future.
Let me know what you think of the approach.

Reply:
You mean an onchain helper type solution that we can make with a single call? I think that would be the ideal solution.

Reply:
I agree that this would be the ideal end goal, but I'd rather start with an offchain solution first (that we can improve/fix if necessary) and move to a helper contract only after it's battle tested.
Since the solution itself would require only 2 onchain calls (like price impact for add/remove liquidity) it should be ok.

Reply:
Yeah, sounds good. So just to be clear - for now, we're going to keep an offchain method on PriceImpact which will support v2 and v3?

Reply:
Yes 👍